orchagent/interface_rates.lua

```diff
@@ -0,0 +1,90 @@
+local one_minus_alpha = 1 - alpha
+local delta = tonumber(ARGV[3])
+
+-- Get speeds
+local speeds = {}
+redis.call('SELECT', asic_db)
+local n = table.getn(KEYS)
+for i = 1, n do
+    speeds[i] = 40000 
+
+redis.call('SELECT', counters_db)
+
+local initialized = redis.call('GET', rates_table_name .. ':' .. 'INIT_DONE')
+
+for i = 1, n do
+    -- Get new COUNTERS values
+    local in_octets = redis.call('HGET', counters_table_name .. ':' .. KEYS[i], 'SAI_ROUTER_INTERFACE_STAT_IN_OCTETS')
+    local in_pkts = redis.call('HGET', counters_table_name .. ':' .. KEYS[i], 'SAI_ROUTER_INTERFACE_STAT_IN_PACKETS')
+    local out_octets = redis.call('HGET', counters_table_name .. ':' .. KEYS[i], 'SAI_ROUTER_INTERFACE_STAT_OUT_OCTETS')
+    local out_pkts = redis.call('HGET', counters_table_name .. ':' .. KEYS[i], 'SAI_ROUTER_INTERFACE_STAT_OUT_PACKETS')
+
+    if initialized == "DONE" or initialized == "COUNTERS_LAST" then
+        -- Get old COUNTERS values
+        local in_octets_pkts_last = redis.call('HGET', rates_table_name .. ':' .. KEYS[i], 'SAI_ROUTER_INTERFACE_STAT_IN_OCTETS_last')
+        local in_pkts_last = redis.call('HGET', rates_table_name .. ':' .. KEYS[i], 'SAI_ROUTER_INTERFACE_STAT_IN_PACKETS_last')
+        local out_octets_last = redis.call('HGET', rates_table_name .. ':' .. KEYS[i], 'SAI_ROUTER_INTERFACE_STAT_OUT_OCTETS_last')
+        local out_pkts_last = redis.call('HGET', rates_table_name .. ':' .. KEYS[i], 'SAI_ROUTER_INTERFACE_STAT_OUT_PACKETS_last')
+        
+        -- Calculate new rates values
+        local rx_bps_new = (in_octets - in_octets_last)/delta
+        local tx_bps_new = (out_octets - out_octets_last)/delta
+        local rx_pps_new = (in_pkts - in_pkts_last)/delta
+        local tx_pps_new = (out_pkts - out_pkts_last)/delta
+        local rx_util_new = rx_bps_new/speeds[i]
+        local tx_util_new = tx_bps_new/speeds[i]
+
+        if initialized == "DONE" then
+            -- Get old rates values
+            local rx_bps_old = redis.call('HGET', rates_table_name .. ':' .. KEYS[i], 'RX_BPS')
+            local rx_pps_old = redis.call('HGET', rates_table_name .. ':' .. KEYS[i], 'RX_PPS')
+            local tx_bps_old = redis.call('HGET', rates_table_name .. ':' .. KEYS[i], 'TX_BPS')
+            local tx_pps_old = redis.call('HGET', rates_table_name .. ':' .. KEYS[i], 'TX_PPS')
+            local rx_util_old = redis.call('HGET', rates_table_name .. ':' .. KEYS[i], 'RX_UTIL')
+            local tx_util_old = redis.call('HGET', rates_table_name .. ':' .. KEYS[i], 'TX_UTIL')
+
+            -- Smooth the rates values and store them in DB
+            redis.call('HSET', rates_table_name .. ':' .. KEYS[i], 'RX_BPS', alpha*rx_bps_new + one_minus_alpha*rx_bps_old)
+            redis.call('HSET', rates_table_name .. ':' .. KEYS[i], 'RX_PPS', alpha*rx_bps_new + one_minus_alpha*rx_bps_old)
+            redis.call('HSET', rates_table_name .. ':' .. KEYS[i], 'TX_BPS', alpha*rx_bps_new + one_minus_alpha*rx_bps_old)
```

Comment:
The logic is including rx_bps_new for all rates type. Is it typo  ?  I assume it should be based on rate type . Also is this Smoothing Logic is verified ?

Reply:
Good catch